### PR TITLE
Changed backup file naming conventions to incorporate content id.

### PR DESCRIPTION
**Pull request by jmcatamney**
Previously, backup files were keyed only to the dbid of the segment from
which they were backed up, which caused problems when trying to restore
files backed up on a mirror segment, restore files to a new cluster with
a different dbid mapping, and so forth.

This commit fixes the above issues by changing the backup file naming
convention to incorporate the content id of the backed-up segment and
enables restoring files based on the content id.  Also, gpdbrestore
seamlessly handles mixing the old and new filename formats, so users can
still restore a backup made with a GPDB 4.3 version of gpcrondump that
used the old format.

### CODE REVIEW

**Comment by xinzweb** [comment]
Thanks @jmcatamney, is that possible to break down this large PR into smaller commits in this PR? It's easier for people to understand this PR. Thanks.

**Comment by jmcatamney** [comment]
@xinzweb It's difficult to split up this commit into logical chunks, as a change to filename conventions touches every single backup- and restore-related file and they all have to be committed together to function correctly, and we squashed this branch a while ago so we don't have any intermediate state of smaller initial commits available.

If it helps, only 11 of the files have the important functional changes--five Python files (gpcrondump, gpdbrestore, dump.py, restore.py, and backup_utils.py) and six C files (cdbbackup.c, cdb_dump.c, cdb_dump_agent.c, cdb_restore.c, cdb_restore_agent.c, and cdb_dump_util.c)--and the rest is testing code, so you could review it in three chunks of Python source, C source, and tests.  I could split it into three commits grouping those files together for review purposes, but I'd rather not if I can avoid it since again none of those commits could function independently.

**Comment by andreasscherbaum** [comment]
How are the changes around getGpArray related to the backup feature? Or is this merely a code redesign?

**Comment by khuddlefish** [comment]
@andreasscherbaum It's a code redesign. There are a couple refactors unrelated to the file-name change in this commit, but since we have lost the commit history from squashing, it would be very difficult to separate them out at this point.

**Comment by jmcatamney** [comment]
@andreasscherbaum The getGpArray changes actually are related to the content id feature, somewhat.  We need to have a mapping between dbids and content ids to generate the filenames, and to do so we need to create a GpArray when creating the Context; since we now have one in Context anyway, we decided to save it and use that wherever else we need a GpArray instead of doing an initFromCatalog in a bunch of different places.

File: src/backend/cdb/cdbbackup.c [hunk 1]
```diff
@@ -1756,50 +1775,86 @@ createBackupDirectory(char *pszPathName)
...
-		/* dump file can start with gp_ and also mpp_ for 2.3 backward compatibility */
-		snprintf(pszRegex, regex_len, "^%s(gp|mpp)_dump_[0-9]+_%d_%s(.gz)?$", DUMP_PREFIX, segid, pszBackupKey);
+		pszRegexNew = (char *) palloc(regex_len + 2);
+		pszRegexOld = (char *) palloc(regex_len);
+		pszRegexMaster = (char *) palloc(regex_len + 2);
 
-		masklen = strlen(pszRegex);
-		mask = (pg_wchar *) palloc((masklen + 1) * sizeof(pg_wchar));
-		wmasklen = pg_mb2wchar_with_len(pszRegex, mask, masklen);
+    if (segid == 1)
```
Comment (jimmyyih) [inline src/backend/cdb/cdbbackup.c]: Should be using tabs instead of spaces for the changes in this file.

File: src/bin/pg_dump/cdb/cdb_dump.c [hunk 2]
```diff
@@ -1152,6 +1154,10 @@ fillInputOptions(int argc, char **argv, InputOptions * pInputOpts)
 				addFileNameParam("schema-file", optarg, pInputOpts);
 				include_everything = false;
 				break;
+			case 22:
+                g_is_old_format = true;
```
Comment (jimmyyih) [inline src/bin/pg_dump/cdb/cdb_dump.c]: Tab instead of space.

File: src/bin/pg_dump/cdb/cdb_dump_agent.c [hunk 3]
```diff
@@ -808,6 +811,9 @@ main(int argc, char **argv)
 				include_everything = false;
 				strncpy(schemaFileName, optarg, sizeof(optarg)+1);
 				break;
+            case 19:
```
Comment (jimmyyih) [inline src/bin/pg_dump/cdb/cdb_dump_agent.c]: Tab instead of space.

File: src/bin/pg_dump/cdb/cdb_restore.c [hunk 4]
```diff
@@ -600,7 +602,10 @@ fillInputOptions(int argc, char **argv, InputOptions * pInputOpts)
 				if (schema_level_file != NULL)
 					free(schema_level_file);
 				break;
-
+            case 20:
```
Comment (jimmyyih) [inline src/bin/pg_dump/cdb/cdb_restore.c]: Tab instead of space.

File: src/bin/pg_dump/cdb/cdb_restore_agent.c [hunk 5]
```diff
@@ -472,6 +475,9 @@ main(int argc, char **argv)
 				ddboost_storage_unit = strdup(optarg);
 				break;
 #endif
+            case 20:
```
Comment (jimmyyih) [inline src/bin/pg_dump/cdb/cdb_restore_agent.c]: Tab instead of space.

File: src/bin/pg_dump/cdb/cdb_dump_agent.c [hunk 6]
```diff
@@ -8276,6 +8282,9 @@ formGenericFilePathName(char *keyword, char *pszBackupDirectory, char *pszBackup
 	int			len;
 	char	   *pszBackupFileName;
 
+	if (g_is_old_format) {
```
Comment (jimmyyih) [inline src/bin/pg_dump/cdb/cdb_dump_agent.c]: Remove the curly brackets.

File: src/bin/pg_dump/cdb/cdb_restore_agent.c [hunk 7]
```diff
@@ -1419,10 +1433,13 @@ static char *formDDBoostFileName(char *pszBackupKey, bool isPostData, char *dd_b
         int     segid;
         int     len = 0;
         char    *pszBackupFileName;
-	char 	*dir_name = "db_dumps";		/* default directory */
+	    char 	*dir_name = "db_dumps";		/* default directory */
 
-        instid = g_role;           		/* dispatch node */
+        instid = g_sourceContentID;
         segid = g_sourceDBID;
+        if (g_is_old_format) {
```
Comment (jimmyyih) [inline src/bin/pg_dump/cdb/cdb_restore_agent.c]: Remove the curly brackets.  If not, you have to newline the opening curly bracket as per styling.

File: gpMgmt/bin/gpdbrestore [hunk 8]
```diff
@@ -408,88 +404,90 @@ class GpdbRestore(Operation):
...
+        matching = sorted(matching, key=lambda x: int(x[-18:-4].strip()))
         if len(matching) == 0:
-            raise ExceptionNoStackTraceNeeded("Could not locate Master database dump files at %s:%s" % (host, path))
-        matching = sorted(matching, key=lambda x: int(x[len(prefix):].strip()))
+            raise ExceptionNoStackTraceNeeded("Could not locate Master database dump files in %s" % root)
         if len(matching) > 1:
             dates_and_times = []
             for match in matching:
-                temp = match[len(prefix):]
+                temp = match[-18:-4]
```
Comment (jimmyyih) [inline gpMgmt/bin/gpdbrestore]: Are these lengths guaranteed?  Seems like high maintenance for later down the road if you want to change something in the filename.

File: gpMgmt/bin/gppylib/operations/backup_utils.py [hunk 9]
```diff
@@ -62,45 +61,114 @@ def __init__(self, values=None):
...
-            timestamp = self.timestamp
+    def generate_prefix(self, filetype, dbid=1, content=None, use_old_format=None):
         format_str =  "%sgp_%s_" % (self.dump_prefix, "%s")
         filename = format_str % (self.filename_dict[filetype][0])
         if "%d" in filename:
-            if dbid == 1:
-                filename = filename % (1, 1)
+            if use_old_format:
+                if dbid == 1:
+                    filename = filename % (1, 1)
```
Comment (jmcatamney) [inline gpMgmt/bin/gppylib/operations/backup_utils.py]: We don't plan to modify the filenames ever again as far as substitutions are concerned (this modification was complex enough), but we can make them use keywords if that would help with clarity.

File: gpMgmt/bin/gppylib/operations/backup_utils.py [hunk 10]
```diff
@@ -62,45 +61,114 @@ def __init__(self, values=None):
...
+        report_file = None
+        using_nbu = (self.netbackup_service_host is not None)
+        if using_nbu:
+            restore_file_with_nbu(self, "report", timestamp=timestamp)
+            report_file = self.generate_filename('report', timestamp=timestamp)
+        else:
+            for dump_dir in dump_dirs:
+                report_file_attempt = self.generate_filename('report', timestamp=timestamp, directory=dump_dir)
+                if os.path.exists(report_file_attempt):
+                    report_file = report_file_attempt
```
Comment (jmcatamney) [inline gpMgmt/bin/gppylib/operations/backup_utils.py]: Good catch, will change.

File: gpMgmt/bin/gppylib/operations/backup_utils.py [hunk 11]
```diff
@@ -62,45 +61,114 @@ def __init__(self, values=None):
...
+        if not report_file:
+            raise Exception("Unable to locate report file for timestamp %s" % timestamp)
+        report_contents = get_lines_from_file(report_file)
+        old_metadata = self.generate_filename("metadata", timestamp=timestamp, use_old_format=True, use_compress=False)
+        old_format = False
```
Comment (jmcatamney) [inline gpMgmt/bin/gppylib/operations/backup_utils.py]: Removed.

File: gpMgmt/bin/gppylib/operations/restore.py [hunk 12]
```diff
@@ -704,36 +700,28 @@ def backup_dir_is_writable(self):
...
-        # status file is not set, cdbbackup.c sets it to SEGMENT_DATA_DIRECTORY if not specified.
+        if gpr_path and status_path:
+            restore_line += " --gp-r=%s" % gpr_path
+            restore_line += " --status=%s" % status_path
+
+        if self.context.use_old_filename_format:
+            restore_line += " --old-format "
 
+        if self.context.dump_prefix:
+            logger.info("Adding --prefix")
```
Comment (jmcatamney) [inline gpMgmt/bin/gppylib/operations/restore.py]: This isn't actually a new addition, the log statement is just getting moved as part of consolidating the various dump_string functions; we'll leave this as-is for now.

File: src/backend/cdb/cdbbackup.c [hunk 13]
```diff
@@ -1095,6 +1108,12 @@ gp_restore_launch__(PG_FUNCTION_ARGS)
 	pszBackupFileName = formBackupFilePathName(pszBackupDirectory, pszBackupKey, is_decompress, postDataSchemaOnly, NULL);
 	pszStatusFileName = formStatusFilePathName(statusDirectory, pszBackupKey, false);
 
+#if 0
```
Comment (jmcatamney) [inline src/backend/cdb/cdbbackup.c]: Not sure why it's there, possibly a leftover from something we tried before.  We'll remove it.

File: src/backend/cdb/cdbbackup.c [hunk 14]
```diff
@@ -1756,50 +1775,86 @@ createBackupDirectory(char *pszPathName)
...
 	{
 		int base_len = 64;
-		int	wmasklen,
-			masklen,
+		int	wmasklenNew, wmasklenOld, wmasklenMaster,
+			masklenNew, masklenOld, masklenMaster,
 			regex_len = base_len + strlen(DUMP_PREFIX) + strlen(pszBackupKey);
-			pg_wchar   *mask;
-		pszRegex = (char *) palloc(regex_len);
+			pg_wchar   *maskNew, *maskOld, *maskMaster;
```
Comment (jmcatamney) [inline src/backend/cdb/cdbbackup.c]: Fixed.

File: src/bin/pg_dump/cdb/cdb_dump_agent.c (hunk 3 above)
Comment (jmcatamney) [inline src/bin/pg_dump/cdb/cdb_dump_agent.c]: Fixed.

**Pull request merged**